File: Dynamic_Programming/problem5/problem5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n <= 1:
            return s

        start = 0
        max_len = 0

        def expand(left: int, right: int) -> None:
            nonlocal start, max_len
            while left >= 0 and right < n and s[left] == s[right]:
                length = right - left + 1
                if length > max_len:
                    max_len = length
                    start = left
                left -= 1
                right += 1

        for i in range(n):
            expand(i, i)
            expand(i, i + 1)

        return s[start : start + max_len]

File: Dynamic_Programming/problem5/problem5.cpp
#include <string>

using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int n = static_cast<int>(s.size());
        if (n <= 1) {
            return s;
        }

        int start = 0;
        int maxLen = 0;

        auto expand = [&](int left, int right) {
            while (left >= 0 && right < n && s[left] == s[right]) {
                int len = right - left + 1;
                if (len > maxLen) {
                    maxLen = len;
                    start = left;
                }
                --left;
                ++right;
            }
        };

        for (int i = 0; i < n; ++i) {
            expand(i, i);
            expand(i, i + 1);
        }

        return s.substr(start, maxLen);
    }
};
